src/pcnt.rs

```diff
@@ -0,0 +1,449 @@
+    fn into(self) -> esp_idf_sys::pcnt_ctrl_mode_t {
+        match self {
+            PcntControlMode::Keep => {
+                esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_HOLD
+            }
+            PcntControlMode::Reverse => {
+                esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_INCREASE
+            }
+            PcntControlMode::Disable => {
+                esp_idf_sys::pcnt_channel_edge_action_t_PCNT_CHANNEL_EDGE_ACTION_DECREASE
+            }
+        }
+    }
+}
+
+bitflags! {
+    #[allow(dead_code)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct PcntEventType: u32 {
+        const THRES_1 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_1;
+        const THRES_0 = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_THRES_0;
+        const L_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_L_LIM;
+        const H_LIM = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_H_LIM;
+        const ZERO = esp_idf_sys::pcnt_evt_type_t_PCNT_EVT_ZERO;
+    }
+}
+
+#[doc = " @brief Pulse Counter configuration for a single channel"]
+#[derive(Debug)]
+pub struct PcntConfig {
+    #[doc = "< Pulse input GPIO number, if you want to use GPIO16, enter pulse_gpio_num = 16, a negative value will be ignored"]
+    pub pulse_pin: Option<PcntPin>,
+    #[doc = "< Control signal input GPIO number, a negative value will be ignored"]
+    pub ctrl_pin: Option<PcntPin>,
+    #[doc = "< PCNT low control mode"]
+    pub lctrl_mode: PcntControlMode,
+    #[doc = "< PCNT high control mode"]
+    pub hctrl_mode: PcntControlMode,
+    #[doc = "< PCNT positive edge count mode"]
+    pub pos_mode: PcntCountMode,
+    #[doc = "< PCNT negative edge count mode"]
+    pub neg_mode: PcntCountMode,
+    #[doc = "< Maximum counter value"]
+    pub counter_h_lim: i16,
+    #[doc = "< Minimum counter value"]
+    pub counter_l_lim: i16,
+    #[doc = "< the PCNT channel"]
+    pub channel: PcntChannel,
+}
+
+impl Into<pcnt_config_t> for &PcntConfig {
+    fn into(self) -> pcnt_config_t {
+        pcnt_config_t {
+            pulse_gpio_num: match &self.pulse_pin {
+                Some(pin) => pin.pin_num(),
+                None => esp_idf_sys::PCNT_PIN_NOT_USED,
+            },
+            ctrl_gpio_num: match &self.ctrl_pin {
+                Some(pin) => pin.pin_num(),
+                None => esp_idf_sys::PCNT_PIN_NOT_USED,
+            },
+            lctrl_mode: self.lctrl_mode.into(),
+            hctrl_mode: self.hctrl_mode.into(),
+            pos_mode: self.pos_mode.into(),
+            neg_mode: self.neg_mode.into(),
+            counter_h_lim: self.counter_h_lim,
+            counter_l_lim: self.counter_l_lim,
+            channel: self.channel.into(),
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PcntPin {
+    pin: Arc<i32>
+}
+
+impl PcntPin {
+    pub fn new<'d>(pin: impl Peripheral<P = impl IOPin> + 'd, pull_mode: Pull) -> Result<Self, EspError>
```

Comment:
I tried to make this `Peripheral<P = impl InputPin>` but when I tried to create the pin driver to set the pull_mode it complains.  I also could not figure out how to set the pin mode before calling this as I'd could not convert the pin driver back to an InputPin implementation.

Reply:
if you need to set the pull mode and later want to get the gpio number you can just store the PinDriver inside your PcntPin like the following
```
pub struct PcntPin<'d> {
    pin: PinDriver<'d, AnyIOPin, Input>
}
```
and inside the constructor you dont need to call .into_input thats redundant afaik. because you already make it an input Driver. it could look like this
```
{
    let pin: PeripheralRef<AnyIOPin> = pin.into_ref().map_into();
    let mut pin = PinDriver.input(pin)?;
    pin.set_pull(pull_mode);
    Ok(Self{pin})
}

fn pin_num(&self) -> i32 {
    self.pin.pin()
}
```

Reply:
PcntPin now stores PinDriver